### PR TITLE
feat: update email service

### PR DESCRIPTION
**Pull request by uyousafzai54**
## Notion ticket link

[Update Email Service](https://www.notion.so/uwblueprintexecs/Tasks-38b49f83f6ce4b07ae188dabb37631e8?p=fc2f2b6bf4b346b9a0f8fc964bf418f7)

## Implementation description
* 

## Steps to test
1. Add following lines to server.ts but replace \<INSERT EMAIL\> with your email.
```
import IEmailService from "./services/interfaces/emailService";
import EmailService from "./services/implementations/emailService";
import nodemailerConfig from "./nodemailer.config";

const emailService: IEmailService = new EmailService(nodemailerConfig);

const email = "<INSERT EMAIL>";
const campers: any = [
  { name: "Anne Chung", age: 22 },
  { name: "Peppa Pig", age: 12 }
];

emailService.sendConfirmationEmail(
  email,
  "Anne Chung",
  "Camp 1",
  "Your mom's house ;)",
  "March 1st, 2022",
  campers,
  "911",
  20,
  40,
  10,
  120,
  "https://www.youtube.com/watch?v=dQw4w9WgXcQ",
);
emailService.sendCancellationConfirmationEmail(email, "Anne Chung");
emailService.sendWaitlistConfirmationEmail(
  email,
  "Anne Chung",
  "Camp 1",
  "Your mom's house ;)",
  "March 1st, 2022",
  campers,
  "911",
);
emailService.sendRegistrationInviteEmail(
  email,
  "Camp 1",
  "Anne Chung",
  "March 1st, 2020",
  "https://www.youtube.com/watch?v=dQw4w9WgXcQ",
);
emailService.sendSpecialNeedsNoticeEmail(
  email,
  "Anne Chung",
  "Camp 1",
  "March 1, 2022",
  "Peppa Pig",
  "annechung301@gmail.com",
  "911",
  "snorting",
);

emailService.sendFullCampNoticeEmail(
  email,
  "Waterloo Camp",
  "March 1st 2022, March 2nd 2022",
);
emailService.sendCamperCancellationNoticeEmail(
  email,
  "Anne Chung",
  "Camp 1",
  "March 1st - March 20th, 2022",
);
```
2. Save the file and start the app.
3. Verify emails have been received and have correct formatting.

## What should reviewers focus on?
* 

## Checklist
- [ ] My PR name is descriptive and in imperative tense
- [ ] My commit messages are descriptive and in imperative tense. My commits are atomic and trivial commits are squashed or fixup'd into non-trivial commits
- [x] I have run the appropriate linter(s)
- [ ] I have requested a review from the PL, as well as other devs who have background knowledge on this PR or who will be building on top of this PR

### CODE REVIEW

File: backend/typescript/services/implementations/emailService.ts [hunk 1]
```diff
@@ -20,31 +20,156 @@ class EmailService implements IEmailService {
...
+    totalCampFees: number,
+    dropOffAndPickupFees: number,
+    totalPayment: number,
+    link: string,
+  ): Promise<void> {
     await this.sendEmail(
       to,
-      "Focus on Nature: Confirmation Email",
-      "dummy body",
+      "Focus on Nature Camp Registration - Confirmation Email",
```
Comment (annechung) [inline backend/typescript/services/implementations/emailService.ts:38]: STOP! look at the file in your code editor pls, the line breaks look so ugly here

File: backend/typescript/services/implementations/emailService.ts [hunk 2]
```diff
@@ -20,31 +20,156 @@ class EmailService implements IEmailService {
...
+      to,
+      "Special Needs Camper Registration Notice",
+      `This following email is to notify you of a special needs camper 
+      registration. The camper listed below checked the special needs box 
+      on their registration form: <br><br>
+      <ul>
+        <li><b>Name of camper:</b> ${camperName} </li>
+        <li><b>Camp name:</b> ${campName} </li>
+        <li><b>Session dates:</b> ${sessionDates} </li>
+        <li><b>Camper's age:</b> ${camperAge} </li>
```
Comment (amanda-guo) [inline backend/typescript/services/implementations/emailService.ts:138]: from the google doc it looks like camper's age is not needed..?

File: backend/typescript/services/implementations/emailService.ts [hunk 3]
```diff
@@ -20,31 +20,156 @@ class EmailService implements IEmailService {
...
+  async sendFullCampNoticeEmail(
+    to: string,
+    campName: string,
+    sessionDates: string,
+  ): Promise<void> {
+    await this.sendEmail(
+      to,
+      "Camp Registration Notice - FULL CAPACITY",
+      `This following email is to notify you that ${campName} is full for the 
+      following session dates: ${sessionDates}`,
```
Comment (amanda-guo) [inline backend/typescript/services/implementations/emailService.ts:157]: nit pick: add a period at the end of the `${sessionDates}`

File: backend/typescript/services/implementations/emailService.ts [hunk 4]
```diff
@@ -20,31 +20,156 @@ class EmailService implements IEmailService {
...
+    sessionDates: string,
+  ): Promise<void> {
     await this.sendEmail(
       to,
-      "Focus on Nature: Cancellation Email",
-      "dummy body",
+      "Camper Cancellation Notice",
+      `This following email is to notify you that ${camperName} has canceled 
+      their camp registration. A spot has now opened up in ${campName} for the 
+      following session dates: ${sessionDates}`,
```
Comment (amanda-guo) [inline backend/typescript/services/implementations/emailService.ts:172]: nit pick: add a period at the end of the `${sessionDates}`

File: backend/typescript/services/implementations/emailService.ts [hunk 5]
```diff
@@ -20,31 +20,156 @@ class EmailService implements IEmailService {
     }
   }
 
-  async sendConfirmationEmail(to: string): Promise<void> {
+  async sendConfirmationEmail(
+    to: string,
+    registrantName: string,
+    campName: string,
+    sessionDates: string,
```
Comment (amanda-guo) [inline backend/typescript/services/implementations/emailService.ts:27]: question: is there a specific format that the session dates should be specified in? i.e. should the admins follow a date-month-year format or is it up to them how they want to input it?

**Review by amanda-guo — COMMENTED** [review]
Works as expected! I was thoroughly entertained from reading the emails LOL

File: backend/typescript/services/implementations/emailService.ts (hunk 5 above)
Comment (annechung) [reply backend/typescript/services/implementations/emailService.ts:28]: according to the email copy it should be [DD, MONTH YEAR - DD, MONTH YEAR] - so sessionDates will have to be passed in in that format

Comment (amanda-guo) [reply backend/typescript/services/implementations/emailService.ts:28]: ok yea makes sense. i guess it'll be enforced on the front end then?

**Review by amanda-guo — APPROVED** [review]
Works as expected and the copy matches w/ the google doc!

File: backend/typescript/services/implementations/emailService.ts (hunk 5 above)
Comment (annechung) [reply backend/typescript/services/implementations/emailService.ts:28]: it'd be wherever these methods were being called - from the backend services

**Pull request merged**